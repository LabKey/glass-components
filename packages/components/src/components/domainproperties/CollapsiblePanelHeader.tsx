import React from "react";
import {Panel} from "react-bootstrap";
import {faCheckCircle, faExclamationCircle, faMinusSquare, faPlusSquare} from "@fortawesome/free-solid-svg-icons";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import { LabelHelpTip } from "../..";
import { DomainPanelStatus } from "./models";

interface Props {
    id: string
    title: string
    titlePrefix?: string
    collapsed: boolean
    collapsible: boolean
    controlledCollapse: boolean
    headerDetails?: string
    iconHelpMsg?: string
    panelStatus: DomainPanelStatus
    togglePanel: (evt: any, collapsed?: boolean) => any
    useTheme: boolean
    isValid: boolean
}

export class CollapsiblePanelHeader extends React.PureComponent<Props, any> {

    getPanelHeaderClass(): string {
        const { collapsed, collapsible, controlledCollapse, useTheme } = this.props;

        let classes = 'domain-panel-header ' + ((collapsible || controlledCollapse) ? 'domain-heading-collapsible' : '');
        classes += (!collapsed ? ' domain-panel-header-expanded' : ' domain-panel-header-collapsed');
        if (!collapsed) {
            classes += (useTheme ? ' labkey-page-nav' : ' domain-panel-header-no-theme');
        }

        return classes;
    }

    getHeaderIconHelpMsg(): string {
        const { isValid, panelStatus, iconHelpMsg } = this.props;

        if (!isValid) {
            return iconHelpMsg;
        }

        if (panelStatus === 'TODO') {
            return 'This section does not contain any user defined fields. You may want to review.'
        }

        return undefined;
    }

    getHeaderIconComponent = () => {
        return (
            <span className={this.getHeaderIconClass()}>
                <FontAwesomeIcon icon={this.getHeaderIcon()}/>
            </span>
        )
    };

    getHeaderIconClass() {
        const { collapsed, isValid, panelStatus } = this.props;
        const classes = 'domain-panel-status-icon';

        if (collapsed) {
            if (isValid && panelStatus === 'COMPLETE') {
                return classes + ' domain-panel-status-icon-green';
            }
            return classes + ' domain-panel-status-icon-blue';
        }

        return classes;
    };

    getHeaderIcon() {
        const { isValid, panelStatus } = this.props;

        if (!isValid || panelStatus === 'TODO') {
            return faExclamationCircle;
        }

        return faCheckCircle;
    };

    renderHeader() {
        const { children, collapsed, titlePrefix, panelStatus, controlledCollapse, collapsible, title, headerDetails } = this.props;
        const iconHelpMsg = ((panelStatus && panelStatus !== 'NONE') ? this.getHeaderIconHelpMsg() : undefined);

        return (
            <>
                {/*Header help icon*/}
                {iconHelpMsg &&
                    <LabelHelpTip title={title} body={() => (iconHelpMsg)} placement="top" iconComponent={this.getHeaderIconComponent}/>
                }
                {panelStatus && panelStatus !== 'NONE' && !iconHelpMsg && this.getHeaderIconComponent()}

                {/*Header name*/}
                <span className={'domain-panel-title'}>{(titlePrefix ? titlePrefix + ' - ' : '') + title}</span>

                {/*Expand/Collapse Icon*/}
                {(controlledCollapse || collapsible) && collapsed &&
                    <span className={'pull-right'}>
                        <FontAwesomeIcon size={'lg'} icon={faPlusSquare} className={"domain-form-expand-btn"}/>
                    </span>
                }
                {(controlledCollapse || collapsible) && !collapsed &&
                    <span className={'pull-right'}>
                        <FontAwesomeIcon size={'lg'} icon={faMinusSquare} className={"domain-form-collapse-btn"}/>
                    </span>
                }

                {/*Help tip*/}
                {children &&
                    <LabelHelpTip customStyle={{verticalAlign: 'top', marginLeft: '5px'}} placement={'top'} title={title} body={() => (children)}/>
                }

                {/*Header details, shown on the right side*/}
                {controlledCollapse && headerDetails &&
                    <span className='domain-panel-header-fields-defined'>{headerDetails}</span>
                }
            </>
        )
    }

    render(){
        const { id, togglePanel } = this.props;

        return(
            <Panel.Heading
                id={id}
                onClick={togglePanel}
                className={this.getPanelHeaderClass()}
            >
                {this.renderHeader()}
            </Panel.Heading>
        );
    }
}
