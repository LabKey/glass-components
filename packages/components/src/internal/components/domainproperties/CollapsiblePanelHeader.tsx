import React from 'react';
import classNames from 'classnames';
import { Panel } from 'react-bootstrap';
import { faCheckCircle, faExclamationCircle, faMinusSquare, faPlusSquare } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { DomainPanelStatus } from './models';
import {LabelHelpTip} from "../base/LabelHelpTip";

interface Props {
    id: string;
    title: string;
    titlePrefix?: string;
    collapsed: boolean;
    collapsible: boolean;
    controlledCollapse: boolean;
    headerDetails?: string;
    todoIconHelpMsg?: string;
    iconHelpMsg?: string;
    panelStatus: DomainPanelStatus;
    togglePanel: (evt: any, collapsed?: boolean) => any;
    useTheme: boolean;
    isValid: boolean;
}

export class CollapsiblePanelHeader extends React.PureComponent<Props, any> {
    getPanelHeaderClass(): string {
        const { collapsed, collapsible, controlledCollapse, useTheme } = this.props;

        return classNames('domain-panel-header', {
            'domain-heading-collapsible': collapsible || controlledCollapse,
            'domain-panel-header-expanded': !collapsed,
            'domain-panel-header-collapsed': collapsed,
            'labkey-page-nav': !collapsed && useTheme,
            'domain-panel-header-no-theme': !collapsed && !useTheme,
        });
    }

    getHeaderIconHelpMsg(): string {
        const { isValid, panelStatus, iconHelpMsg, todoIconHelpMsg } = this.props;

        if (!isValid) {
            return iconHelpMsg;
        }

        if (panelStatus === 'TODO') {
            return todoIconHelpMsg || 'This section does not contain any user-defined fields. You may want to review.';
        }

        return undefined;
    }

    getHeaderIconComponent = () => {
        return (
            <span className={this.getHeaderIconClass()}>
                <FontAwesomeIcon icon={this.getHeaderIcon()} />
            </span>
        );
    };

    getHeaderIconClass() {
        const { collapsed, isValid, panelStatus } = this.props;
        const validComplete = isValid && panelStatus === 'COMPLETE';

        return classNames('domain-panel-status-icon', {
            'domain-panel-status-icon-green': collapsed && validComplete,
            'domain-panel-status-icon-blue': collapsed && !validComplete,
        });
    }

    getHeaderIcon() {
        const { isValid, panelStatus } = this.props;
        return !isValid || panelStatus === 'TODO' ? faExclamationCircle : faCheckCircle;
    }

    renderExpandCollapseIcon() {
        const { collapsed } = this.props;
        const icon = collapsed ? faPlusSquare : faMinusSquare;
        const className = collapsed ? 'domain-form-expand-btn' : 'domain-form-collapse-btn';

        return (
            <span className="pull-right">
                <FontAwesomeIcon size="lg" icon={icon} className={className} />
            </span>
        );
    }

    getTitlePrefix(): string {
        let prefix = this.props.titlePrefix;

        // ellipsis after certain length
        if (prefix && prefix.length > 70) {
            prefix = prefix.substr(0, 70) + '...';
        }

        return prefix ? prefix + ' - ' : '';
    }

    renderHeader() {
        const { children, panelStatus, controlledCollapse, collapsible, title, headerDetails } = this.props;
        const iconHelpMsg = panelStatus && panelStatus !== 'NONE' ? this.getHeaderIconHelpMsg() : undefined;

        return (
            <>
                {/* Header help icon*/}
                {iconHelpMsg && (
                    <LabelHelpTip iconComponent={this.getHeaderIconComponent()} placement="top" title={title}>
                        {iconHelpMsg}
                    </LabelHelpTip>
                )}
                {panelStatus && panelStatus !== 'NONE' && !iconHelpMsg && this.getHeaderIconComponent()}

                {/* Header name*/}
                <span className="domain-panel-title">{this.getTitlePrefix() + title}</span>

                {/* Expand/Collapse Icon*/}
                {(controlledCollapse || collapsible) && this.renderExpandCollapseIcon()}

                {/* Help tip*/}
                {children && (
                    <LabelHelpTip
                        customStyle={{ verticalAlign: 'top', marginLeft: '5px' }}
                        placement="top"
                        title={title}
                    >
                        {children}
                    </LabelHelpTip>
                )}

                {/* Header details, shown on the right side*/}
                {controlledCollapse && headerDetails && (
                    <span className="domain-panel-header-fields-defined">{headerDetails}</span>
                )}
            </>
        );
    }

    render() {
        const { id, togglePanel } = this.props;

        return (
            <Panel.Heading id={id} onClick={togglePanel} className={this.getPanelHeaderClass()}>
                {this.renderHeader()}
            </Panel.Heading>
        );
    }
}
