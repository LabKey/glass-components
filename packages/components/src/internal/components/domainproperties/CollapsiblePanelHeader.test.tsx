import React from 'react';
import { render } from '@testing-library/react';

import { CollapsiblePanelHeader } from './CollapsiblePanelHeader';

describe('<CollapsiblePanelHeader/>', () => {
    test('default properties', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                collapsed={true}
                collapsible={true}
                controlledCollapse={true}
                panelStatus="NONE"
                isValid={true}
                togglePanel={jest.fn()}
            />
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });

    test('custom properties', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                collapsed={true}
                collapsible={true}
                controlledCollapse={true}
                isValid={true}
                togglePanel={jest.fn()}
                titlePrefix="Test Prefix"
                headerDetails="N Fields"
                panelStatus="COMPLETE"
            />
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });

    test('not controlledCollapse', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                isValid
                togglePanel={jest.fn()}
                collapsed={true}
                collapsible={false}
                controlledCollapse={false}
                headerDetails="N Fields"
                panelStatus="COMPLETE"
            />
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });

    test('not collapsible', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                panelStatus="NONE"
                isValid
                togglePanel={jest.fn()}
                collapsed={false}
                collapsible={false}
                controlledCollapse={false}
            />
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });

    test('invalid, iconHelpMsg, and expanded', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                collapsed={false}
                collapsible
                controlledCollapse
                togglePanel={jest.fn()}
                isValid
                panelStatus="INPROGRESS"
                iconHelpMsg="Test icon help message"
            />
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });

    test('panelStatus TODO and help tip', () => {
        const component = (
            <CollapsiblePanelHeader
                id="test-id"
                title="Test Title"
                collapsed
                collapsible
                controlledCollapse
                isValid
                togglePanel={jest.fn()}
                panelStatus="TODO"
                todoIconHelpMsg="Some other TODO message goes here."
            >
                <div>Test help tip message</div>
            </CollapsiblePanelHeader>
        );

        const { container } = render(component);
        expect(container).toMatchSnapshot();
    });
});
