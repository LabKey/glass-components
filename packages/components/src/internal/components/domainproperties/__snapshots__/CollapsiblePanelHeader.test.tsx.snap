// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollapsiblePanelHeader/> custom properties 1`] = `
<div>
  <div
    class="domain-panel-header domain-heading-collapsible domain-panel-header-collapsed"
    id="test-id"
  >
    <span
      class="domain-panel-status-icon domain-panel-status-icon-green"
    >
      <span
        class="fa fa-check-circle"
      />
    </span>
    <span
      class="domain-panel-title"
    >
      Test Prefix - Test Title
    </span>
    <span
      class="pull-right"
    >
      <span
        class="fa fa-lg fa-plus-square domain-form-expand-btn"
      />
    </span>
    <span
      class="domain-panel-header-fields-defined"
    >
      N Fields
    </span>
  </div>
</div>
`;

exports[`<CollapsiblePanelHeader/> default properties 1`] = `
<div>
  <div
    class="domain-panel-header domain-heading-collapsible domain-panel-header-collapsed"
    id="test-id"
  >
    <span
      class="domain-panel-title"
    >
      Test Title
    </span>
    <span
      class="pull-right"
    >
      <span
        class="fa fa-lg fa-plus-square domain-form-expand-btn"
      />
    </span>
  </div>
</div>
`;

exports[`<CollapsiblePanelHeader/> invalid, iconHelpMsg, and expanded 1`] = `
<div>
  <div
    class="domain-panel-header domain-heading-collapsible domain-panel-header-expanded labkey-page-nav"
    id="test-id"
  >
    <span
      class="domain-panel-status-icon"
    >
      <span
        class="fa fa-check-circle"
      />
    </span>
    <span
      class="domain-panel-title"
    >
      Test Title
    </span>
    <span
      class="pull-right"
    >
      <span
        class="fa fa-lg fa-minus-square domain-form-collapse-btn"
      />
    </span>
  </div>
</div>
`;

exports[`<CollapsiblePanelHeader/> not collapsible 1`] = `
<div>
  <div
    class="domain-panel-header domain-panel-header-expanded labkey-page-nav"
    id="test-id"
  >
    <span
      class="domain-panel-title"
    >
      Test Title
    </span>
  </div>
</div>
`;

exports[`<CollapsiblePanelHeader/> not controlledCollapse 1`] = `
<div>
  <div
    class="domain-panel-header domain-panel-header-collapsed"
    id="test-id"
  >
    <span
      class="domain-panel-status-icon domain-panel-status-icon-green"
    >
      <span
        class="fa fa-check-circle"
      />
    </span>
    <span
      class="domain-panel-title"
    >
      Test Title
    </span>
  </div>
</div>
`;

exports[`<CollapsiblePanelHeader/> panelStatus TODO and help tip 1`] = `
<div>
  <div
    class="domain-panel-header domain-heading-collapsible domain-panel-header-collapsed"
    id="test-id"
  >
    <span
      class="label-help-target"
    >
      <span
        class="domain-panel-status-icon domain-panel-status-icon-blue"
      >
        <span
          class="fa fa-exclamation-circle"
        />
      </span>
    </span>
    <span
      class="domain-panel-title"
    >
      Test Title
    </span>
    <span
      class="pull-right"
    >
      <span
        class="fa fa-lg fa-plus-square domain-form-expand-btn"
      />
    </span>
    <span
      class="label-help-target"
    >
      <span
        class="label-help-icon fa fa-question-circle"
      />
    </span>
  </div>
</div>
`;
